### PR TITLE
koji-hub crd

### PR DESCRIPTION
**Pull request by odra**
Issue: #10 

This PR enables the operator to deploy koji-hub in kubernetes.

## Dependencies

* postgresql 
* rabbitmq

##  Verification

Running tests locally (any of those suffice, they accomplish the same):
* `operator-sdk test local`
* `molecule test -s test-local`

Manual Deployment:

NOTE: enable ingress addon in case you are using minikube: `minikube addons enable ingress`

* Change the image in mbox-operator/deploy/operator.yaml if you want to use a different operator image;
* deploy postgresql: `kubectl apply -f ../components/psql`
* cd into `mbox-operator` and run `make deploy`;
* Wait for the mbox-operaor pod to be ready
* Deploy koji-hub by creating a CR (custom resource): `k apply -f deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml`
* Wait for the pod to be running
* You can get its ingress host/ip address and test it with curl: `curl -X POST -H 'Content-Type: text:/xml' -d '<?xml version="1.0" encoding="utf-8"?><methodCall><methodName>system.listMethods</methodName><params></params></methodCall>'   http://$(kubectl get ingress/koji-hub-http -o jsonpath='{.status.loadBalancer.ingress[0].ip}')/kojihub/`

### CODE REVIEW

**Comment by odra** [comment]
ping @Zlopez - sending the early pr as we talked in IRC, will update it as I add more stuff

File: mbox-operator/deploy/role.yaml [hunk 1]
```diff
@@ -4,77 +4,78 @@ metadata:
...
-      - '*'
-    verbs:
-      - create
-      - delete
-      - get
-      - list
-      - patch
-      - update
-      - watch
+- apiGroups:
```
Comment (Zlopez) [inline mbox-operator/deploy/role.yaml]: Why this indentation change?

**Comment by Zlopez** [comment]
@odra This looks really nice, thanks for sharing

File: mbox-operator/deploy/role.yaml (hunk 1 above)
Comment (odra) [reply mbox-operator/deploy/role.yaml]: auto generated by the sdk when I added the new crd file

**Comment by Zlopez** [comment]
fedmsg should go away and we should use fedora messaging. There is a RabbitMQ server running in Vagrant, but I'm not sure if koji has fedora messaging support.

**Comment by odra** [comment]
Oh, I added it because there was fedmsg dependencies in kojihub components folder, should I ignore it for now?

**Comment by Zlopez** [comment]
Left it for now, we could remove it in future.

File: mbox-operator/roles/koji-hub/templates/configmap.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    LoginCreateUser = on
+
+    NotifyOnSuccess = False
+
+    Plugins = fedmsg-koji-plugin runroot_hub
+  kojiweb.conf: |-
+    [web]
+    SiteName = MBOX Koji
+    KojiHubURL = https://koji-hub:8443/kojihub
+    KojiFilesURL = https://{{ hostname }}/pkgs
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/configmap.yml.j2]: Shouldn't there be:
```suggestion
    KojiFilesURL = https://{{ koji_hub_hostname }}/pkgs
```

File: mbox-operator/roles/koji-hub/templates/pvc.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/pvc.yml.j2]: This will be shared by the koji-hub and koji-builder, just like in current mbbox.

Comment (odra) [reply mbox-operator/roles/koji-hub/templates/pvc.yml.j2]: why do we need to sare the pv?

Comment (Zlopez) [reply mbox-operator/roles/koji-hub/templates/pvc.yml.j2]: They are using the same database, so it's better to use one pv for both.

File: mbox-operator/molecule/default/asserts/koji-hub.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+              api_version: v1
+              kind: Secret
+              namespace: "{{ namespace }}"
+              name: "{{ kojihub_cacert_secret }}"
+            register: kojihub_secrets
+          - assert:
+              that:
+                - kojihub_secrets.resources|length == 1                
+                - kojihub_secrets.resources[0].metadata.labels['app'] == 'koji-hub'
+                - "'ca.crt' in kojihub_secrets.resources[0].data"
```
Comment (Zlopez) [inline mbox-operator/molecule/default/asserts/koji-hub.yml:39]:
```suggestion
                - "'cert' in kojihub_secrets.resources[0].data"
```
In `koji-hub` role main this is fixed `cert` value.

File: images/koji-hub/service.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: postgresql
```
Comment (Zlopez) [inline images/koji-hub/service.yaml:4]: If this is related to postgres app, why it's part of the koji-hub images?

File: mbox-operator/deploy/cluster_role_binding.yaml [hunk 6]
```diff
@@ -1,27 +0,0 @@
-apiVersion: rbac.authorization.k8s.io/v1
```
Comment (Zlopez) [inline mbox-operator/deploy/cluster_role_binding.yaml:1]: Why the cluster_role_binding is removed?

File: components/psql/0-secret.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
```
Comment (Zlopez) [inline components/psql/0-secret.yaml:1]: This is the postgresql db that was shared by koji-builder and koji-hub in original script?

Also I'm not sure if we want to add anything new to `components` folder. After conversion to operator we will probably remove it or do you think, we should use it for anything?

File: mbox-operator/deploy/operator.yaml [hunk 8]
```diff
@@ -14,21 +14,9 @@ spec:
...
-          image: quay.io/fedora/mbox-operator:latest
-          imagePullPolicy: Always
-          volumeMounts:
-            - mountPath: /tmp/ansible-operator/runner
-              name: runner
-              readOnly: true
-        - name: operator
-          # Replace this with the built image name
-          image: quay.io/fedora/mbox-operator:latest
+          image: quay.io/lrossett/mbox-operator:latest
```
Comment (Zlopez) [inline mbox-operator/deploy/operator.yaml:19]: We should use the `fedora` username, I already uploaded koji-builder image to it.

File: mbox-operator/deploy/operator.yaml.j2 [hunk 9]
```diff
@@ -14,22 +14,10 @@ spec:
     spec:
       serviceAccountName: mbox-operator
       containers:
-        - name: ansible
```
Comment (Zlopez) [inline mbox-operator/deploy/operator.yaml.j2:17]: Why we have `operator.yaml` and jinja template with same name?

Is one of them used for tests and the other for actual deployment?

File: mbox-operator/deploy/role_binding.yaml [hunk 10]
```diff
@@ -8,4 +8,4 @@ subjects:
 roleRef:
   kind: Role
   name: mbox-operator
-  apiGroup: rbac.authorization.k8s.io
+  apiGroup: rbac.authorization.k8s.io
```
Comment (Zlopez) [inline mbox-operator/deploy/role_binding.yaml:11]: Why removing the newline? Or is this some automatic formatting?

File: mbox-operator/molecule/test-cluster/molecule.yml [hunk 11]
```diff
@@ -19,7 +19,7 @@ provisioner:
       all:
         namespace: ${TEST_NAMESPACE:-osdk-test}
         tmpdir: /tmp/mbox
-        operator_image: quay.io/fedora/mbox-operator:latest
+        operator_image: quay.io/lrossett/mbox-operator:latest
```
Comment (Zlopez) [inline mbox-operator/molecule/test-cluster/molecule.yml:22]: We should use the `fedora` namespace.

File: mbox-operator/molecule/test-cluster/molecule.yml [hunk 12]
```diff
@@ -32,13 +32,12 @@ scenario:
...
     - syntax
-    # add a create playbook to create crs
-    # - create
     - prepare
+    - create
     - converge
     - side_effect
     - verify
-    - destroy
+    #- destroy
```
Comment (Zlopez) [inline mbox-operator/molecule/test-cluster/molecule.yml:40]: Why destroy is commented out?

File: mbox-operator/molecule/test-cluster/prepare.yml [hunk 13]
```diff
@@ -27,12 +27,6 @@
         - role_binding.yaml
         - service_account.yaml
 
-    - name: Create cluster RBAC resources
```
Comment (Zlopez) [inline mbox-operator/molecule/test-cluster/prepare.yml:30]: To be honest, I'm not sure what exactly RBAC means or why it is in the original template.

File: mbox-operator/molecule/test-local/prepare.yml [hunk 14]
```diff
@@ -27,3 +30,4 @@
         - role.yaml
         - role_binding.yaml
         - service_account.yaml
+
```
Comment (Zlopez) [inline mbox-operator/molecule/test-local/prepare.yml:33]: Why this empty line?

File: mbox-operator/roles/koji-hub/defaults/main.yml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+---
+# defaults file for koji-hub
+koji_hub_image: quay.io/fedora/koji-hub:latest
+koji_hub_replicas: 1
+
+koji_hub_configmap: koji-hub-configmap
+koji_hub_fedmsg_configmap: fedmsg-configmap
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/defaults/main.yml:7]: Is this still needed?

File: mbox-operator/roles/koji-hub/templates/service.yml.j2 [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  name: koji-hub
+  labels:
+    app: koji-hub
+    service: koji-hub
+spec:
+  # type: NodePort
+  ports:
+    - name: http
+      port: 80
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/service.yml.j2:12]: Original script was using port 8443, why this change?

File: mbox-operator/roles/koji-hub/templates/deployment.yml.j2 [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+      labels:
+        app: koji-hub
+    spec:
+      containers:
+        - name: koji-hub
+          image: "{{ koji_hub_image }}"
+          readinessProbe:
+            exec:
+              command:
+                - /readiness.sh
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/deployment.yml.j2:23]: Is this some kubernetes script? I can find it in the PR itself.

File: mbox-operator/roles/koji-hub/templates/pvc.yml.j2 (hunk 3 above)
Comment (odra) [reply mbox-operator/roles/koji-hub/templates/pvc.yml.j2:1]: it's an external postgres right? they shouldn't share a pvc if that's the only reason to use a shared pv.

File: mbox-operator/molecule/test-cluster/prepare.yml (hunk 13 above)
Comment (odra) [reply mbox-operator/molecule/test-cluster/prepare.yml:30]: RBAC = Role Based Access Control, it is relayed to resources that are related to role role bindings and service accounts: https://kubernetes.io/docs/reference/access-authn-authz/rbac/

File: mbox-operator/molecule/test-local/prepare.yml (hunk 14 above)
Comment (odra) [reply mbox-operator/molecule/test-local/prepare.yml:33]: bad habit :)

File: mbox-operator/roles/koji-hub/templates/service.yml.j2 (hunk 16 above)
Comment (odra) [reply mbox-operator/roles/koji-hub/templates/service.yml.j2:12]: well its unsafe so 80, I couold change it to 8080 and 8443.

File: mbox-operator/roles/koji-hub/templates/deployment.yml.j2 (hunk 17 above)
Comment (odra) [reply mbox-operator/roles/koji-hub/templates/deployment.yml.j2:23]: The script lies within the koji-hub container.

I added it as a script because readiness probes cannot execute post requests which are mandatory for XML RPC calls.

File: mbox-operator/roles/koji-hub/templates/ingress.yml.j2 [hunk 18]
```diff
@@ -11,4 +11,4 @@ metadata:
 spec:
   backend:
     serviceName: koji-hub
-    servicePort: {{ koji_hub_target_port }}
+    servicePort: 80
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/ingress.yml.j2:14]: Shouldn't this be 8080?

File: mbox-operator/roles/koji-hub/templates/deployment.yml.j2 (hunk 17 above)
Comment (Zlopez) [reply mbox-operator/roles/koji-hub/templates/deployment.yml.j2:27]: I didn't noticed this script before in Dockerfile, now it makes sense.

File: mbox-operator/molecule/test-cluster/prepare.yml (hunk 13 above)
Comment (Zlopez) [reply mbox-operator/molecule/test-cluster/prepare.yml:30]: Thanks for clarification. Do you think we don't need this kind of access control?

File: mbox-operator/deploy/cluster_role_binding.yaml (hunk 6 above)
Comment (Zlopez) [reply mbox-operator/deploy/cluster_role_binding.yaml:1]: I understand this now, when I found out what RBAC is.

File: mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+  persistent: true
+  hostname: ''
+  configmap: koji-hub
+  cacert_secret: cacert
+  service_cert_secret: service-cert
+  postgres_secret: postgres
+  http_enabled: true
+  https_enabled: true
+  topic_prefix: mbox_dev
+  fedora_messaging_url: amqps://koji:something@rabbitmq_something.fedoraproject.org/%2Fpubsub
```
Comment (Zlopez) [inline mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:18]: Just to be sure, these variables are the ones that user will be able to configure when preparing deployment?

File: mbox-operator/roles/koji-hub/defaults/main.yml [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+koji_hub_httpd_pvc_name: koji-hub-httpd-pvc
+koji_hub_httpd_pvc_size: 1Gi
+
+koji_hub_mnt_pvc_name: koji-hub-mnt-pvc
+koji_hub_mnt_pvc_size: 10Gi
+
+koji_hub_http_port: "8080"
+koji_hub_https_port: "8443"
+koji_hub_svc_name: koji-hub
+koji_hub_http_enabled: "{{ http_enabled|default(true) }}"
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/defaults/main.yml:20]: This should be false by default. It doesn't make sense to have both http and https enabled by default.

File: mbox-operator/roles/koji-hub/templates/deployment.yml.j2 [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+    metadata:
+      labels:
+        app: koji-hub
+    spec:
+      containers:
+        - name: koji-hub
+          image: "{{ koji_hub_image }}"
+          livenessProbe:
+            tcpSocket:
+              port: {{ koji_hub_http_port }}
```
Comment (Zlopez) [inline mbox-operator/roles/koji-hub/templates/deployment.yml.j2:22]: Shouldn't there be condition for HTTPS port?

File: mbox-operator/roles/koji-hub/templates/pvc.yml.j2 (hunk 3 above)
Comment (Zlopez) [reply mbox-operator/roles/koji-hub/templates/pvc.yml.j2:1]: Yes, the original solution created a standalone postgres db and the koji pods just connected to this one.

File: mbox-operator/roles/koji-hub/templates/deployment.yml.j2 (hunk 21 above)
Comment (odra) [reply mbox-operator/roles/koji-hub/templates/deployment.yml.j2:22]: yeah I am doing a few changes for that

File: mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml (hunk 19 above)
Comment (odra) [reply mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:18]: yes, those are available in the operator as global vars.

File: mbox-operator/molecule/test-cluster/prepare.yml (hunk 13 above)
Comment (odra) [reply mbox-operator/molecule/test-cluster/prepare.yml:30]: nope, I had an issue in minikube but we don't need a cluster wide perm.

Comment (Zlopez) [reply mbox-operator/molecule/test-cluster/prepare.yml:30]: Ok, I would remove it also from koji-builder

File: mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml (hunk 19 above)
Comment (Zlopez) [reply mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:18]: Thanks, I need to look at the koji-builder what should be configurable than

File: mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml [hunk 22]
```diff
@@ -0,0 +1,18 @@
+apiVersion: apps.fedoraproject.org/v1alpha1
+kind: MBKojiHub
+metadata:
+  name: mb-koji-hub
+  labels:
+    app: mb-koji-hub
+spec:
+  replicas: 1
+  persistent: true
+  host: '192.168.39.143'
```
Comment (Zlopez) [inline mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:10]: Why the hardcoded IP?

File: mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+  configmap: koji-hub
+  cacert_secret: cacert
+  service_cert_secret: service-cert
+  postgres_secret: postgres
+  http_enabled: true
+  https_enabled: true
```
Comment (Zlopez) [inline mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:16]: How does work, that both of these options are enabled? Does HTTPS has priority in this case?

Comment (odra) [reply mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:16]: it would support each one individually as if it using different ports.

Comment (Zlopez) [reply mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:16]: I think we should disable HTTP by default.

**Comment by Zlopez** [comment]
I tried to test it locally and I found that vagrant is missing a few things and created #29. The tests were successful and everything looks fine.

**Comment by odra** [comment]
I just need push a minor change for mq support and it should  be good to go

**Comment by odra** [comment]
ok done

File: components/rabbitmq/0-secret.yaml [hunk 24]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
```
Comment (Zlopez) [inline components/rabbitmq/0-secret.yaml:1]: I assume the rabbitmq is only for testing purposes and it's not part of the mbox deployment. At least it should be optional. The same should be true for psql.

Comment (odra) [reply components/rabbitmq/0-secret.yaml:1]: yeah it's just for development environment purposes.
